File: cml/src/main/java/com/fiwio/iot/demeter/DemeterIoInteractor.java
package com.fiwio.iot.demeter;

import com.fiwio.iot.demeter.device.model.DigitalIO;
import com.fiwio.iot.demeter.device.model.DigitalPins;
import com.fiwio.iot.demeter.device.model.DigitalValue;
import com.fiwio.iot.demeter.fsm.IOInteractor;

import au.com.ds.ef.StatefulContext;

class DemeterIoInteractor implements IOInteractor {
  private final DigitalPins demeter;
  private final DigitalIO barrel_input;
  private final DigitalIO barrel_pump;
  private final DigitalIO branch_valve;

  public DemeterIoInteractor(DigitalPins demeter, String branchVentilName) {
    this.demeter = demeter;
    barrel_input = demeter.getInput("INP0");
    barrel_pump = demeter.getOutput("BCM23");
    branch_valve = demeter.getOutput(branchVentilName);
  }

  @Override
  public boolean swimmerIsInactive() {
    return barrel_input.getValue() == DigitalValue.ON;
  }

  @Override
  public void barrelPumpOn(StatefulContext context) {

    barrel_pump.setValue(DigitalValue.ON);
  }

  @Override
  public void openBarrel(StatefulContext context) {
    branch_valve.setValue(DigitalValue.ON);
  }

  @Override
  public void closeAllVentils(StatefulContext context) {
    barrel_pump.setValue(DigitalValue.OFF);
    branch_valve.setValue(DigitalValue.OFF);
  }
}

File: domain/src/main/java/com/fiwio/iot/demeter/domain/features/fsm/DemeterIOInteractor.kt
package com.fiwio.iot.demeter.domain.features.fsm

import au.com.ds.ef.StatefulContext
import com.fiwio.iot.demeter.domain.features.manual.SwitchActuator
import com.fiwio.iot.demeter.domain.features.tracking.EventTracker
import com.fiwio.iot.demeter.domain.model.Actuator
import com.fiwio.iot.demeter.domain.model.Demeter
import com.fiwio.iot.demeter.domain.model.InputValue
import com.fiwio.iot.demeter.domain.model.Sensor
import io.reactivex.observers.DisposableSingleObserver

class DemeterIOInteractor(val barrel_pump: Actuator, val barrel_input: Sensor, val barrel_valve: Actuator,
                          val switchActuator: SwitchActuator, val tracker: EventTracker) : IOInteractor {

    override fun swimmerIsInactive(): Boolean {
        return (barrel_input.value === InputValue.ON);
    }

    override fun openBarrel(ctx: StatefulContext) {
        switchActuator.execute(object : DisposableSingleObserver<Demeter>() {
            override fun onSuccess(demeter: Demeter) {
                tracker.track("barrel was opened:" + ctx.id)
            }

            override fun onError(e: Throwable) {
                tracker.track("error by barrel opening;" + ctx.id)
            }
        }, Actuator(barrel_valve.name, true))
    }

    override fun barrelPumpOn(ctx: StatefulContext) {
        switchActuator.execute(object : DisposableSingleObserver<Demeter>() {
            override fun onSuccess(demeter: Demeter) {
                tracker.track("pump was started:" + ctx.id)
            }

            override fun onError(e: Throwable) {
                tracker.track("error by pump start:" + ctx.id)
            }
        }, Actuator(barrel_pump.name, true))
    }

    // TODO create batch command for settings multiple actuator at once
    override fun closeAllVentils(ctx: StatefulContext) {
        switchActuator.execute(object : DisposableSingleObserver<Demeter>() {
            override fun onSuccess(demeter: Demeter) {
                tracker.track("barrel was closed:" + ctx.id)
            }

            override fun onError(e: Throwable) {
                tracker.track("error by brrel closing:" + ctx.id)

            }
        }, Actuator(barrel_valve.name, false))

        switchActuator.execute(object : DisposableSingleObserver<Demeter>() {
            override fun onSuccess(demeter: Demeter) {
                tracker.track("pump was stoped:" + ctx.id)
            }

            override fun onError(e: Throwable) {
                tracker.track("error by stoping the pump:" + ctx.id)

            }
        }, Actuator(barrel_pump.name, false))

    }
}